src/store/index-level.ts

```diff
@@ -1,213 +1,393 @@
+   * @param options IndexLevelOptions that include an AbortSignal.
    */
   async put(
     tenant: string,
-    dataId: string,
-    indexes: { [property: string]: unknown },
+    itemId: string,
+    indexes: KeyValues,
     options?: IndexLevelOptions
   ): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
-    indexes = flatten(indexes);
 
-    const operations: LevelWrapperBatchOperation<string>[] = [ ];
+    // ensure we have something valid to index, remove any empty strings from index properties
+    removeEmptyStrings(indexes);
+    if (isEmptyObject(indexes)) {
+      throw new DwnError(DwnErrorCode.IndexMissingIndexableProperty, 'Index must include at least one valid indexable property');
+    }
 
-    // create an index entry for each property in the `indexes`
-    for (const propertyName in indexes) {
-      const propertyValue = indexes[propertyName];
+    const tenantPartition = await this.db.partition(tenant);
+    const indexOps: LevelWrapperBatchOperation<string>[] = [];
 
-      // NOTE: appending data ID after (property + value) serves two purposes:
-      // 1. creates a unique entry of the property-value pair per data/object
-      // 2. when we need to delete all indexes of a given data ID (`delete()`), we can reconstruct the index keys and remove the indexes efficiently
+    // create an index entry for each property
+    // these indexes are all sortable lexicographically.
+    for (const sortProperty in indexes) {
+      const sortValue = indexes[sortProperty];
+      // the key is sortValue followed by the itemId as a tie-breaker.
+      // for example if the property is messageTimestamp the key would look like:
+      //  '"2023-05-25T18:23:29.425008Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
       //
-      // example keys (\u0000 is just shown for illustration purpose because it is the delimiter used to join the string segments below):
-      // 'interface\u0000"Records"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'method\u0000"Write"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'schema\u0000"http://ud4kyzon6ugxn64boz7v"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'dataCid\u0000"bafkreic3ie3cxsblp46vn3ofumdnwiqqk4d5ah7uqgpcn6xps4skfvagze"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'dateCreated\u0000"2023-05-25T18:23:29.425008Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      const key = this.join(propertyName, this.encodeValue(propertyValue), dataId);
-      operations.push({ type: 'put', key, value: dataId });
+      const key = IndexLevel.keySegmentJoin(IndexLevel.encodeValue(sortValue), itemId);
+      const item: IndexedItem = { itemId, indexes };
```

Comment:
Guess I will find out as I read on why indexes  need to be included for every index write/push. Immediate reaction assuming I am reading it correctly is that this seems rather heavy. e.g. for every `RecordsWrite`, we could be looking at close to 20 key-value pairs stored repetitively ~20 times.